xarray/core/duck_array_ops.py

```diff
@@ -219,22 +219,10 @@ def asarray(data, xp=np):
 
 
 def as_shared_dtype(scalars_or_arrays, xp=np):
-    """Cast a arrays to a shared dtype using xarray's type promotion rules."""
-    array_type_cupy = array_type("cupy")
-    if array_type_cupy and any(
-        isinstance(x, array_type_cupy) for x in scalars_or_arrays
-    ):
-        import cupy as cp
-
-        arrays = [asarray(x, xp=cp) for x in scalars_or_arrays]
-    else:
-        arrays = [asarray(x, xp=xp) for x in scalars_or_arrays]
-    # Pass arrays directly instead of dtypes to result_type so scalars
-    # get handled properly.
-    # Note that result_type() safely gets the dtype from dask arrays without
-    # evaluating them.
-    out_type = dtypes.result_type(*arrays)
-    return [astype(x, out_type, copy=False) for x in arrays]
+    """Cast arrays to a shared dtype using xarray's type promotion rules."""
+    duckarrays = [to_duck_array(obj, xp=xp) for obj in scalars_or_arrays]
```

Comment:
When you say "read from disk" do you meaning calling the `__array__` attribute of the innermost duckarray? Because that's what I'm trying to avoid.

Reply:
> do you meaning calling the __array__ attribute of the innermost duckarray?

I think our naming convention is that "duck array" is a "computational array" e.g.numpy, dask but NOT our explicitly-indexed array classes. The latter wrap duck arrays.

Read from disk should be happening by calling `get_duck_array` on the outermost ExplicitlyIndexed class, which should propagate down to `BackendArray` which reads bytes using either indexing or `np.asarray` (I think).

(related : https://github.com/zarr-developers/zarr-python/discussions/1603#discussioncomment-8171930)

PS: We could chat on a call some time if you want. It's all quite confusing :) This is a good opportunity to add some comments/docs for devs

Reply:
> Read from disk should be happening by calling get_duck_array on the outermost ExplicitlyIndexed class, which should propagate down to BackendArray which reads bytes using either indexing or np.asarray (I think).

Yes, but my `KerchunkArray` case is interesting because I don't want to use `BackendArray` (I have no use for CopyOnWrite because I'm never loading bytes, nor for Lazy indexing (I can't index into the `KerchunkArray` at all).

> PS: We could chat on a call some time if you want. It's all quite confusing :) This is a good opportunity to add some comments/docs for devs

Yeah that could be helpful actually :) I'm learning a lot right now about a part of xarray I have never had a reason to look at before!

Reply:
> but my KerchunkArray case is interesting because I don't want to use BackendArray

Well then don't use the backend infrastructure? :P

Reply:
Haha yes yes 

No but seriously I did think about that and I do think that it does make sense to use the backend infrastructure here. I could make my full case, but after all we are still reading from files here, we just aren't reading the bytes inside the chunks.